src/test/FreeC/Pass/KindCheckPassTests.hs

```diff
@@ -0,0 +1,182 @@
+      _ <- defineTestTypeCon "State" 2
+      _ <- defineTestTypeCon "Int" 0
+      checkType input
+  it "should not accept underapplied type synonyms" $ do
+    input <- expectParseTestType "State Int"
+    shouldFail $ do
+      _ <- defineTestTypeSyn "State" ["s", "a"] "s -> (,) s a"
+      _ <- defineTestTypeCon "Int" 0
+      _ <- defineTestTypeCon "(,)" 2
+      checkType input
+  it "should not accept overapplied type constructors" $ do
+    input <- expectParseTestType "State Int Int Int"
+    shouldFail $ do
+      _ <- defineTestTypeCon "State" 2
+      _ <- defineTestTypeCon "Int" 0
+      checkType input
+  it "should not accept overapplied type synonyms" $ do
+    input <- expectParseTestType "State Int Int Int"
+    shouldFail $ do
+      _ <- defineTestTypeSyn "State" ["s", "a"] "s -> (,) s a"
+      _ <- defineTestTypeCon "Int" 0
+      _ <- defineTestTypeCon "(,)" 2
+      checkType input
+  it "should not accept type variable applications in function type signatures"
+    $ do
+        input <- expectParseTestModule
+          ["module M where", "f :: forall m a. m a -> m a;", "f x = x;"]
+        shouldFail $ do
+          _ <- defineTestTypeVar "m"
+          _ <- defineTestTypeVar "a"
+          _ <- defineTestVar "x"
+          _ <- defineTestFunc "f" 1 "forall a. m a -> m a"
+          kindCheckPass input
+  it "should not accept type variable applications in function return types"
+    $ do
+        input <- expectParseTestModule ["module M where", "f x :: m a = x;"]
+        shouldFail $ do
+          _ <- defineTestTypeVar "m"
+          _ <- defineTestTypeVar "a"
+          _ <- defineTestVar "x"
+          _ <- defineTestFunc "f" 1 "forall m a. m a -> m a"
+          kindCheckPass input
+  it
+      "should not accept type variable applications in type annotated function arguments"
+    $ do
+        input <- expectParseTestModule ["module M where", "f (x :: m a) = x;"]
+        shouldFail $ do
+          _ <- defineTestTypeVar "m"
+          _ <- defineTestTypeVar "a"
+          _ <- defineTestVar "x"
+          _ <- defineTestFunc "f" 1 "forall m a. m a -> m a"
+          kindCheckPass input
+  it "should not accept type variable applications in type annotated variables"
+    $ do
+        input <- expectParseTestModule ["module M where", "f x = x :: m a;"]
+        shouldFail $ do
+          _ <- defineTestTypeVar "m"
+          _ <- defineTestTypeVar "a"
+          _ <- defineTestVar "x"
+          _ <- defineTestFunc "f" 1 "forall m a. m a -> m a"
+          kindCheckPass input
+  it
+      "should not accept type variable applications in type annotated case expression variables"
+    $ do
+        input <- expectParseTestModule
+          ["module M where", "f x = case x of {C (y :: m b) -> y};"]
+        shouldFail $ do
+          _ <- defineTestTypeVar "m"
+          _ <- defineTestTypeVar "a"
+          _ <- defineTestTypeVar "b"
+          _ <- defineTestVar "x"
+          _ <- defineTestVar "y"
```

Comment:
When you have a lot of variables to introduce, it might be still readable enough when you map over all variables you like to introduce?

```suggestion
          _ <- mapM defineTestTypeVar ["m", "a", "b"]
          _ <- mapM defineTestVar ["x","y"]
```